File: backend/src/main/java/com/localcloud/photos/utils/HashUtils.java
package com.localcloud.photos.utils;

import java.io.InputStream;
import java.security.MessageDigest;

public class HashUtils {

    public static String calculateSHA256(InputStream is) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] byteArray = new byte[1024];
            int bytesCount = 0;

            while ((bytesCount = is.read(byteArray)) != -1) {
                digest.update(byteArray, 0, bytesCount);
            }

            byte[] bytes = digest.digest();

            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < bytes.length; i++) {
                sb.append(Integer.toString((bytes[i] & 0xff) + 0x100, 16).substring(1));
            }

            return sb.toString();
        } catch (Exception e) {
            throw new RuntimeException("Unable to calculate hash", e);
        }
    }
}

File: client/app/src/main/java/com/localcloud/photosclient/utils/HashUtils.kt
package com.localcloud.photosclient.utils

import java.io.File
import java.io.FileInputStream
import java.security.MessageDigest

object HashUtils {
    fun calculateSHA256(inputStream: java.io.InputStream): String {
        val digest = MessageDigest.getInstance("SHA-256")
        inputStream.use { fis ->
            val byteArray = ByteArray(1024)
            var bytesCount: Int
            while (fis.read(byteArray).also { bytesCount = it } != -1) {
                digest.update(byteArray, 0, bytesCount)
            }
        }
        val bytes = digest.digest()
        val sb = java.lang.StringBuilder()
        for (i in bytes.indices) {
            sb.append(Integer.toString((bytes[i].toInt() and 0xff) + 0x100, 16).substring(1))
        }
        return sb.toString()
    }
}
